aiokatcp/client.py

```diff
@@ -529,3 +561,358 @@ def _set_last_exc(self, future: asyncio.Future, exc: Exception) -> None:
+    """
+
+    def __init__(self, client: Client) -> None:
+        self.client = client
+        self.logger = client.logger
+        client.add_connected_callback(self._connected)
+        client.add_disconnected_callback(self._disconnected)
+        client.add_inform_callback('interface-changed', self._interface_changed)
+        client.add_inform_callback('sensor-status', self._sensor_status)
+        self._update_task = None           # type: Optional[asyncio.Task]
+        # Sensors we have seen: maps name to arguments
+        self._sensors = {}                 # type: Dict[str, Tuple[bytes, ...]]
+        # Sensors whose sampling strategy has been set
+        self._sampling_set = set()         # type: Set[str]
+        self._in_batch = False
+        self._watchers = set()             # type: Set[SensorWatcher]
+
+    def add_watcher(self, watcher: SensorWatcher) -> None:
+        self._watchers.add(watcher)
+
+    def remove_watcher(self, watcher: SensorWatcher) -> None:
+        self._watchers.discard(watcher)
+
+    def __bool__(self) -> bool:
+        """True if there are any watchers"""
+        return bool(self._watchers)
+
+    @contextlib.contextmanager
+    def _batch(self) -> Generator[None, None, None]:
+        assert not self._in_batch, 'Re-entered _batch'
+        self._in_batch = True
+        self.logger.debug('Entering batch')
+        try:
+            for watcher in self._watchers:
+                watcher.batch_start()
+            yield
+        finally:
+            self.logger.debug('Exiting batch')
+            self._in_batch = False
+            for watcher in self._watchers:
+                watcher.batch_stop()
+
+    def _cancel_update(self) -> None:
+        if self._update_task is not None:
+            self._update_task.cancel()
+            self._update_task = None
+
+    def _update_done(self, future):
+        try:
+            future.result()
+        except asyncio.CancelledError:
+            pass
+        except OSError as error:
+            # Connection died before we finished. Log it, but no need for
+            # a stack trace.
+            self.logger.warning('Connection error in update task: %s', error)
+        except Exception:
+            self.logger.exception('Exception in update task')
+
+    def _trigger_update(self) -> None:
+        self.logger.debug('Sensor sync triggered')
+        for watcher in self._watchers:
+            watcher.state_updated(SyncState.UNSYNCED)
+        self._cancel_update()
+        self._update_task = self.client.loop.create_task(self._update())
+        self._update_task.add_done_callback(self._update_done)
+
+    async def _set_sampling(self, names: Sequence[str]) -> None:
+        """Register sampling strategy with sensors in `names`"""
+        coros = [self.client.request('sensor-sampling', name, 'auto') for name in names]
+        results = await asyncio.gather(*coros, loop=self.client.loop, return_exceptions=True)
```

Comment:
Some KATCP servers might not like this burst of requests.  And some implementations (e.g. the MeerKAT CMC and DMC) don't handle requests concurrently - complaining if they get a new request from the same client before the previous request is finished.

Reply:
From what I remember, the CMC/DMC also don't implement message IDs. When aiokatcp.Client encounters such a server, it serialises requests.

If a server reads more messages from the socket than it can cope with and chokes as a result, I'd consider it to be a bug in the server or the server's katcp implementation. But if we need to work around such servers, I think the place to do it is in the implementation of `Client.request` rather than here.

It will be interesting to see how well this does at subscribing to a large number of sensors compared to the bulk subscription extension.